torch/_inductor/codegen/cpp.py

```diff
@@ -1043,15 +1188,83 @@ def reduction(name, dtype, src_dtype, reduction_type, index, value):
 
             @staticmethod
             def constant(val, dtype):
+                current_node: torch.fx.Node = V.interpreter.current_node
+                current_node.meta["dtype"] = dtype
+                i32_iinfo = numpy.iinfo(numpy.int32)
+                if (
+                    dtype == torch.int64
+                    and val <= i32_iinfo.max
+                    and val >= i32_iinfo.min
+                ):
+                    current_node.meta["dtype"] = torch.int32
+
+                f32_iinfo = numpy.finfo(numpy.float32)
+                if dtype == torch.double:
+                    if (
+                        (val <= f32_iinfo.max and val >= f32_iinfo.min)
+                        or (val == numpy.inf)
+                        or (val == -numpy.inf)
+                    ):
+                        current_node.meta["dtype"] = torch.float32
+
                 supported_dtype = (torch.float32, torch.int32)
-                is_supported_dtype = dtype in (supported_dtype)
+                is_supported_dtype = current_node.meta["dtype"] in (supported_dtype)
                 if not is_supported_dtype:
                     self.simd_vec = False
                 return is_supported_dtype
 
             @staticmethod
             def index_expr(expr, dtype):
```

Comment:
The logic is a bit complicated. Give a brief description on what is supported in the docstring?

Reply:
I submitted another PR on top of this ghstack. Do you think it could address your comment?

Reply:
Sounds good.